pyomo/contrib/pynumero/sparse/mpi_block_vector.py

```diff
@@ -0,0 +1,1291 @@
+
+    def _comparison_helper(self, other, operation):
+        assert_block_structure(self)
+        result = self.copy_structure()
+        if isinstance(other, MPIBlockVector):
+            assert_block_structure(other)
+            assert self.nblocks == other.nblocks, \
+                'Number of blocks mismatch: {} != {}'.format(self.nblocks, other.nblocks)
+            assert np.array_equal(self._rank_owner, other._rank_owner), \
+                'MPIBlockVectors must be distributed in same processors'
+            assert self._mpiw == other._mpiw, 'Need to have same communicator'
+
+            for i in self._owned_blocks:
+                result.set_block(i, operation(self.get_block(i), other.get_block(i)))
+            return result
+        elif isinstance(other, BlockVector):
+            raise RuntimeError('Operation not supported by MPIBlockVector')
+        elif isinstance(other, np.ndarray):
+            raise RuntimeError('Operation not supported by MPIBlockVector')
+        elif np.isscalar(other):
+            for i in self._owned_blocks:
+                result.set_block(i, operation(self.get_block(i), other))
+            return result
+        else:
+            raise NotImplementedError('Operation not supported by MPIBlockVector')
+
+    def __le__(self, other):
+        return self._comparison_helper(other, operator.le)
+
+    def __lt__(self, other):
+        return self._comparison_helper(other, operator.lt)
+
+    def __ge__(self, other):
+        return self._comparison_helper(other, operator.ge)
+
+    def __gt__(self, other):
+        return self._comparison_helper(other, operator.gt)
+
+    def __eq__(self, other):
+        return self._comparison_helper(other, operator.eq)
+
+    def __ne__(self, other):
+        return self._comparison_helper(other, operator.ne)
+
+    def __contains__(self, item):
+        other = item
+        assert_block_structure(self)
+        if np.isscalar(other):
+            contains = False
+            for i in self._owned_blocks:
+                if other in self.get_block(i):
+                    contains = True
+            return bool(self._mpiw.allreduce(contains, op=MPI.SUM))
+        else:
+            raise NotImplementedError('Operation not supported by MPIBlockVector')
+
+    def get_block(self, key):
+        owner = self._rank_owner[key]
+        rank = self._mpiw.Get_rank()
+        assert owner == rank or owner < 0, 'Block {} not own by processor {}'.format(key, rank)
+        return self._block_vector.get_block(key)
+
+    def set_block(self, key, value):
+        owner = self._rank_owner[key]
+        rank = self._mpiw.Get_rank()
+        assert owner == rank or owner < 0, \
+            'Block {} not owned by processor {}'.format(key, rank)
+
+        self._block_vector.set_block(key, value)
+        self._set_block_size(key, value.size)
+
+    def __getitem__(self, item):
+        raise NotImplementedError('MPIBlockVector does not support __getitem__.')
```

Comment:
Should we point the user to get_block and set_block in these error messages?

Reply:
Done.